File: PT07Y.py
from collections import defaultdict


n, m = tuple(map(int, input().split()))
if m >= n:
    print("NO")
else:
    nodes = defaultdict(lambda: False)
    for _ in range(m):
        a, b = tuple(map(int, input().split()))
        if nodes[a] and nodes[b]:
            print("NO")
            break
        nodes[a] = nodes[b] = True
    else:
        print("NO" if len(nodes) < n else "YES")

File: PT07Y.lua
n, m = io.read("*number", "*number")
if m ~= n - 1 then
    print("NO")
else
    nodes = {}
    for _ = 1, m do
        a, b = io.read("*number", "*number")
        if nodes[a] and nodes[b] then
            print("NO")
            os.exit()
        end
        nodes[a], nodes[b] = true, true
    end
    if #nodes ~= n then
        print("NO")
    else
        print("YES")
    end
end
